### PR TITLE
core: remove refcount for bpf program

### LINKED ISSUE #19270: bpf: deprecate user space refcount
**Is your feature request related to a problem? Please describe.**
Opening an issue for @keszybz proposal https://github.com/systemd/systemd/pull/13496#pullrequestreview-632304667 for steps to simplify the existing BPF infrastructure.

**Describe the solution you'd like**
1. Stop tracking ref count of bpf programs in user space, since kernel already does that, deprecate `bpf_program_ref`, `bpf_program_unref` helpers.
2.  Introduce (`attach_type`, `fd`) tuple as a handle identifying  a program loaded to the kernel. Simplify `struct BPFProgram`, so it stores only bpf instructions related fields and doesn't have key semantics.
3.  Apply the above to BPF features: bpf-firewall, bpf-devices, bpf-foreign.

### PR DESCRIPTION
**Pull request by alexlzhu**
kernel already implements this. Thus the ref count we keep for
bpf-program is redundant.

This PR removes ref count for bpf program as part of a task to simplify
bpf-program and remove redundancies, which will make the switch to
code-compiled BPF programs easier.

Part of #19270

### CODE REVIEW

**Comment by bluca** [comment]
Did the kernel always implement this, or is it a recent addition?

**Comment by alexlzhu** [comment]
Please refer to this article for refcount in the kernel. https://facebookmicrosites.github.io/bpf/blog/2018/08/31/object-lifetime.html. It has been there at least three years.

**Comment by keszybz** [comment]
Please split the commit description into a one-line title (not more than ~60 characters), an empty line, and the rest of the description.

File: src/shared/bpf-program.c [hunk 1]
```diff
@@ -38,7 +38,28 @@ static const char *const bpf_cgroup_attach_type_table[__MAX_BPF_ATTACH_TYPE] = {
...
 DEFINE_STRING_TABLE_LOOKUP(bpf_cgroup_attach_type, int);
 
-DEFINE_HASH_OPS_WITH_KEY_DESTRUCTOR(bpf_program_hash_ops, void, trivial_hash_func, trivial_compare_func, bpf_program_unref);
+
+BPFProgram *bpf_program_free(BPFProgram *p) {
+        if (!p)
+            return NULL;
+        /* Unfortunately, the kernel currently doesn't implicitly detach BPF programs from their cgroups when the last
+         * fd to the BPF program is closed. This has nasty side-effects since this means that abnormally terminated
+         * programs that attached one of their BPF programs to a cgroup will leave this programs pinned for good with
```
Comment (keszybz) [inline src/shared/bpf-program.c]: "this program"

**Comment by bluca** [comment]
> Please refer to this article for refcount in the kernel. https://facebookmicrosites.github.io/bpf/blog/2018/08/31/object-lifetime.html. It has been there at least three years.

Sorry but I don't really have time to delve into long blog posts, and "3 years" doesn't really answer the question. Does every kernel that supports BPF also do refcounting implicitly?

File: src/core/bpf-firewall.c [hunk 2]
```diff
@@ -651,31 +651,24 @@ int bpf_firewall_load_custom(Unit *u) {
...
 static int attach_custom_bpf_progs(Unit *u, const char *path, int attach_type, Set **set, Set **set_installed) {
-        BPFProgram *prog;
+        _cleanup_(bpf_program_freep) BPFProgram *prog;
         int r;
 
         assert(u);
 
-        set_clear(*set_installed);
+        set_clear_free(*set_installed);
+        *set_installed = set_new(NULL);
```
Comment (anitazha) [inline src/core/bpf-firewall.c]: This is leaking. You need to define a set with the hash op so it uses the proper free function. You want `set_new(&bpf_program_hash_ops)` instead

**Comment by alexlzhu** [comment]
Also, this PR is not currently ready for review. Sorry for any misunderstanding.

File: src/shared/bpf-program.c [hunk 3]
```diff
@@ -38,7 +38,28 @@ static const char *const bpf_cgroup_attach_type_table[__MAX_BPF_ATTACH_TYPE] = {
 
 DEFINE_STRING_TABLE_LOOKUP(bpf_cgroup_attach_type, int);
 
-DEFINE_HASH_OPS_WITH_KEY_DESTRUCTOR(bpf_program_hash_ops, void, trivial_hash_func, trivial_compare_func, bpf_program_unref);
+
+BPFProgram *bpf_program_free(BPFProgram *p) {
```
Comment (jkartseva) [inline src/shared/bpf-program.c:43]: nit: I see that the intent of moving this code block is to place `bpf_program_free` before `bpf_program_hash_ops` but that is not needed since `bpf_program_free` is declared in header file. Code will perfectly compile w/o this change.
If you still want to move, please do it in a separate commit to sort out refactoring an functional changes

File: src/shared/bpf-program.c [hunk 4]
```diff
@@ -410,7 +408,7 @@ int bpf_program_serialize_attachment(
 }
 
 int bpf_program_serialize_attachment_set(FILE *f, FDSet *fds, const char *key, Set *set) {
-        BPFProgram *p;
+        _cleanup_(bpf_program_freep) BPFProgram *p;
```
Comment (jkartseva) [inline src/shared/bpf-program.c]: Ouch, this line introduces two bugs:

1. If `set` is empty freeing uninitialized pointer is undefined behavior, so please make sure that ` _cleanup_` always applied to a pointer which is initialized, e.g.:
```
_cleanup_(foo_freep) Foo *p = NULL;
```
2. It `set` is not empty last pointer assigned to `p` will be double freed: the first time when `p` goes out of scope, then when `Set *set` is cleaned up (and the pointers in it are freed).

Summarizing, please don't change this line. `_cleanup_` is required when we transfer pointer's ownership (imagine `unique_ptr` in C++), what we do here is simply read-only iterate.

File: src/shared/bpf-program.c [hunk 5]
```diff
@@ -470,22 +468,21 @@ int bpf_program_deserialize_attachment(const char *v, FDSet *fds, BPFProgram **b
...
-                bpf_program_unref(*bpfp);
+                bpf_program_free(*bpfp);
 
         *bpfp = TAKE_PTR(p);
         return 0;
 }
 
 int bpf_program_deserialize_attachment_set(const char *v, FDSet *fds, Set **bpfsetp) {
-        BPFProgram *p = NULL;
+        _cleanup_(bpf_program_freep) BPFProgram *p = NULL;
```
Comment (jkartseva) [inline src/shared/bpf-program.c]: Similarly to `bpf_program_serialize_attachment_set`, introducing `_cleanup_` will lead to double-free, first time when `p` is out of scope, then when `Set **bpfsetp` is cleaned up.

To prevent that from happening:
1. Don't introduce this change
2. Alternatively, add TAKE_PTR(p) after `set_ensure_consume`:
```
        r = bpf_program_deserialize_attachment(v, fds, &p);
        if (r < 0)
                return r;

        TAKE_PTR(p);
```

File: src/core/bpf-firewall.c [hunk 6]
```diff
@@ -651,31 +651,24 @@ int bpf_firewall_load_custom(Unit *u) {
 }
 
 static int attach_custom_bpf_progs(Unit *u, const char *path, int attach_type, Set **set, Set **set_installed) {
-        BPFProgram *prog;
+        _cleanup_(bpf_program_freep) BPFProgram *prog;
```
Comment (jkartseva) [inline src/core/bpf-firewall.c]: Please also see a comment for `bpf_program_serialize_attachment_set`.

When `*set` is empty, `prog` remains uninitialized: it's undefined behavior.
Also `_cleanup_` introduces double-free, please leave it as `BPFProgram *prog;`

File: src/core/bpf-firewall.c [hunk 7]
```diff
@@ -651,31 +651,24 @@ int bpf_firewall_load_custom(Unit *u) {
...
         int r;
 
         assert(u);
 
-        set_clear(*set_installed);
+        set_clear_free(*set_installed);
+        *set_installed = set_new(&bpf_program_hash_ops);
```
Comment (jkartseva) [inline src/core/bpf-firewall.c]: `set_clear_free` followed by `set_new` will result in an empty set: essentially the same that it is with `set_clear`.
`*set_installed` will store a different address though. What's the point of that?

File: src/core/cgroup.c [hunk 8]
```diff
@@ -1164,7 +1164,7 @@ static void cgroup_apply_restrict_network_interfaces(Unit *u) {
 }
 
 static int cgroup_apply_devices(Unit *u) {
-        _cleanup_(bpf_program_unrefp) BPFProgram *prog = NULL;
+        BPFProgram *prog = NULL;
```
Comment (jkartseva) [inline src/core/cgroup.c]: This will cause memory leak on the error path in between `bpf_devices_cgroup_init` (when `prog` is initialized) and `bpf_devices_apply_policy` (when `prog` is assigned to `bpf_device_control_installed`) calls.
Luckily, there is no `return` in between... sill this change has this potential danger.

What about leaving `_cleanup_`:
```
_cleanup_(bpf_program_freep) BPFProgram *prog = NULL;`
```

Modifying `bpf_devices_apply_policy` signature, so it accepts pointer-to-pointer:

```
bpf_devices_apply_policy(
        BPFProgram **prog,
         /*...*/
``` 

`TAKE_PTR` when assigning `prog` to `bpf_device_control_installed`:

```
 *prog_installed = TAKE_PTR(*prog);
```

?

**Review by wat-ze-hex — CHANGES_REQUESTED** [review]
@alexlzhu 

Thank you for working on that! I think removing `n_ref` from `BPFProgram` is a solid first step towards simplification (and potentially deprecation) of pre-libbpf code base. 

There are a few places which need rework, please see the inline comments.

**Comment by wat-ze-hex** [comment]
@bluca 

> Did the kernel always implement this, or is it a recent addition?

Yes

> Sorry but I don't really have time to delve into long blog posts, and "3 years" doesn't really answer the question. Does every kernel that supports BPF also do refcounting implicitly?

Yes

refcnt has being supported since the dawn of ebpf, i.e. 3.18. 
Please take a look at `struct bpf_prog_aux` [refcnt field](https://elixir.bootlin.com/linux/v3.18.140/source/include/linux/bpf.h#L123) also at [kernel/bpf/syscall.c](https://elixir.bootlin.com/linux/v3.18.140/source/kernel/bpf/syscall.c#L445). refcnt is increased by `bpf_prog_get` and reset by `bpf_prog_put`.

File: src/core/bpf-firewall.c (hunk 7 above)
Comment (alexlzhu) [reply src/core/bpf-firewall.c]: `bpf_program_hash_ops` is needed to specify the free function needed to free the `bpf_program` objects from the set. Previously, that was specified in `set_ensure_put`, but since here we instead use `SET_FOREACH_MOVE` we no longer can specify `bpf_program_hash_ops` except through `set_new`. It does appear that `set_clear` is sufficient though.

File: src/core/bpf-devices.c [hunk 9]
```diff
@@ -219,7 +219,7 @@ int bpf_devices_apply_policy(
 
         /* This will assign *prog_installed if everything goes well. */
 
-        if (!prog)
+        if (!*prog)
```
Comment (anitazha) [inline src/core/bpf-devices.c:223]: Should assert(prog) before using it

File: src/core/bpf-firewall.c [hunk 10]
```diff
@@ -657,25 +657,18 @@ static int attach_custom_bpf_progs(Unit *u, const char *path, int attach_type, S
         assert(u);
 
         set_clear(*set_installed);
+        *set_installed = set_new(&bpf_program_hash_ops);
```
Comment (anitazha) [inline src/core/bpf-firewall.c]: I think you want `set_ensure_allocated()` instead of `set_new()`. The former checks if the set is already allocated. `set_clear()` doesn't free the set so the old `set_installed` could be leaked when you overwrite it with this new one.

**Review by anitazha — CHANGES_REQUESTED** [review]

**Comment by anitazha** [comment]
I think this is on track to get merged so please publish it to take it out of draft.

Also update the PR and commit message so as not to close #19270 (maybe write "Part of #19270"). This PR only handles one of the items on the list.

**Comment by alexlzhu** [comment]
This PR should be ready for review now. Thanks!

**Comment by bluca** [comment]
> @bluca
> 
> > Did the kernel always implement this, or is it a recent addition?
> 
> Yes
> 
> > Sorry but I don't really have time to delve into long blog posts, and "3 years" doesn't really answer the question. Does every kernel that supports BPF also do refcounting implicitly?
> 
> Yes
> 
> refcnt has been supported since the dawn of ebpf, i.e. 3.18. Please take a look at `struct bpf_prog_aux` [refcnt field](https://elixir.bootlin.com/linux/v3.18.140/source/include/linux/bpf.h#L123) also at [kernel/bpf/syscall.c](https://elixir.bootlin.com/linux/v3.18.140/source/kernel/bpf/syscall.c#L445). refcnt is increased by `bpf_prog_get` and reset by `bpf_prog_put`.

Thank you, exactly what I was looking for, that's great

File: src/test/test-bpf-devices.c [hunk 11]
```diff
@@ -15,7 +15,7 @@
 #include "tests.h"
 
 static void test_policy_closed(const char *cgroup_path, BPFProgram **installed_prog) {
-        _cleanup_(bpf_program_unrefp) BPFProgram *prog = NULL;
+        BPFProgram *prog = NULL;
```
Comment (jkartseva) [inline src/test/test-bpf-devices.c]: I didn't grok this in the prior review: what frees `prog` pointer after `_cleanup_` is removed? Looks like memleak. Please refer to `int cgroup_apply_devices(Unit *u);` where the corresponding pointer is wrapped in `_cleanup_`.

File: src/test/test-bpf-devices.c [hunk 12]
```diff
@@ -53,7 +53,7 @@ static void test_policy_closed(const char *cgroup_path, BPFProgram **installed_p
 }
 
 static void test_policy_strict(const char *cgroup_path, BPFProgram **installed_prog) {
-        _cleanup_(bpf_program_unrefp) BPFProgram *prog = NULL;
+        BPFProgram *prog = NULL;
```
Comment (jkartseva) [inline src/test/test-bpf-devices.c]: ditto here and below

File: src/test/test-bpf-devices.c (hunk 11 above)
Comment (alexlzhu) [reply src/test/test-bpf-devices.c]: It is rather odd, I do not see memory leaks in valgrind with or without the free here. I'll add it back in.

Comment (jkartseva) [reply src/test/test-bpf-devices.c]: @alexlzhu

Perhaps they are skipped and don't get to this point.

Comment (alexlzhu) [reply src/test/test-bpf-devices.c]: They are not skipped, I believe it is because of TAKE_PTR in 'bpf_devices_apply_policy' from the previous iteration.

File: src/core/bpf-firewall.c [hunk 13]
```diff
@@ -657,25 +657,18 @@ static int attach_custom_bpf_progs(Unit *u, const char *path, int attach_type, S
         assert(u);
 
         set_clear(*set_installed);
+        set_ensure_allocated(set_installed, &bpf_program_hash_ops);
```
Comment (keszybz) [inline src/core/bpf-firewall.c:660]: Coverity complains that the return value is not checked here.

Comment (yuwata) [reply src/core/bpf-firewall.c:660]: Fix is waiting in #20997.
